tests/test_websockets.py

```diff
@@ -0,0 +1,150 @@
+    async def handler(request, ws, **params):
+        assert ws.subprotocol is None
+        ev.set()
+
+    with liveclient as query:
+        response = await query('GET', '/ws', headers={
+            'Upgrade': 'websocket',
+            'Connection': 'upgrade',
+            'Sec-WebSocket-Key': 'hojIvDoHedBucveephosh8==',
+            'Sec-WebSocket-Version': '13'})
+
+    assert ev.is_set()
+    assert response.status == 101
+
+    # With keep-alive in Connection
+    with liveclient as query:
+        response = await query('GET', '/ws', headers={
+            'Upgrade': 'websocket',
+            'Connection': 'keep-alive, upgrade',
+            'Sec-WebSocket-Key': 'hojIvDoHedBucveephosh8==',
+            'Sec-WebSocket-Version': '13'})
+
+    assert response.status == 101
+
+
+@pytest.mark.asyncio
+async def test_websocket_communication(liveclient):
+
+    @liveclient.app.route('/echo', protocol="websocket")
+    async def echo(request, ws, **params):
+        while True:
+            message = await ws.recv()
+            await ws.send(message)
+
+    # Echo
+    websocket = await websockets.connect(liveclient.wsl + '/echo')
+    try:
+        for message in ('This', 'is', 'a', 'simple', 'test'):
+            await websocket.send(message)
+            echoed = await websocket.recv()
+            assert echoed == message
+    finally:
+        await websocket.close()
+
+    # Nothing sent, just close
+    websocket = await websockets.connect(liveclient.wsl + '/echo')
+    await websocket.close()
+
+
+@pytest.mark.asyncio
+async def test_websocket_broadcasting(liveclient):
+
+    @liveclient.app.route('/broadcast', protocol="websocket")
+    async def feed(request, ws, **params):
+        while True:
+            message = await ws.recv()
+            await asyncio.wait([
+                socket.send(message) for (task, socket)
+                in request.app['websockets'] if socket != ws])
+
+    # connecting
+    connected = []
+    for n in range(1, 6):
+        ws = await websockets.connect(liveclient.wsl + '/broadcast')
+        connected.append(ws)
+
+    # Broadcasting
+    for wid, ws in enumerate(connected, 1):
+        broadcast = 'this is a broadcast from {}'.format(wid)
+        await ws.send(broadcast)
+        for ows in connected:
+            if ows != ws:
+                message = await ows.recv()
+                assert message == broadcast
+
+    # Closing
+    for ws in connected:
+        await ws.close()
+
+
+@pytest.mark.asyncio
+async def test_websocket_binary(liveclient):
+
+    @liveclient.app.route('/bin', protocol="websocket")
+    async def binary(request, ws, **params):
+        await ws.send(b'test')
+
+    # Echo
+    websocket = await websockets.connect(liveclient.wsl + '/bin')
+    bdata = await websocket.recv()
+    await websocket.close_connection_task
+    assert bdata == b'test'
+    assert websocket.close_reason == ''
+    assert websocket.state == 3
```

Comment:
Should be `State.CLOSED`, same for other tests.

Reply:
Done